packages/expo-image/ios/EXImageView.m

```diff
@@ -5,21 +5,154 @@
+
+@end
+
 @implementation EXImageView
 
 - (void)dealloc
 {
   // Stop any active operations or downloads
-  [self sd_setImageWithURL:nil];
+  [self sd_cancelCurrentImageLoad];
 }
 
 # pragma mark -  Custom prop setters
 
 - (void)setSource:(NSDictionary *)source
 {
-  NSURL *imageURL = [RCTConvert NSURL:source[sourceUriKey]];
+  _imageURL = [RCTConvert NSURL:source[sourceUriKey]];
+}
+
+- (void)didSetProps:(NSArray<NSString *> *)changedProps
+{
+  if ([changedProps containsObject:@"source"]) {
+    [self updateImage];
+  }
+}
+
+- (void)updateImage
+{
+  // We want to call onError, onLoadEnd for the previous image load
+  // before calling onLoadStart for the next image load.
+  [self sd_cancelCurrentImageLoad];
+
+  if (self.onLoadStart) {
+    self.onLoadStart(@{
+      @"url": _imageURL.absoluteString ?: [NSNull null]
+    });
+  }
+
+  [self sd_setImageWithURL:_imageURL
+          placeholderImage:nil
+                   options:0
+                  progress:[self progressBlock]
+                 completed:[self completionBlock]];
+}
 
-  [self sd_setImageWithURL:imageURL];
+- (SDImageLoaderProgressBlock)progressBlock
+{
+  __weak EXImageView *weakSelf = self;
+  return ^(NSInteger receivedSize, NSInteger expectedSize, NSURL * _Nullable targetURL) {
+    __strong EXImageView *strongSelf = weakSelf;
+    if (strongSelf.onProgress) {
+      strongSelf.onProgress(@{
+        @"loaded": @(receivedSize),
+        @"total": @(expectedSize)
+      });
+    }
+  };
+}
+
+- (SDExternalCompletionBlock)completionBlock
+{
+  __weak EXImageView *weakSelf = self;
+  return ^(UIImage * _Nullable image, NSError * _Nullable error, SDImageCacheType cacheType, NSURL * _Nullable imageURL) {
+    __strong EXImageView *strongSelf = weakSelf;
+    if (!strongSelf) {
+      // Nothing to do
+      return;
+    }
+
+    if (error && strongSelf.onError) {
+      strongSelf.onError(@{
+        @"error": error.localizedDescription,
+        @"ios": @{
+            @"code": @(error.code),
+            @"domain": error.domain,
+            @"description": error.localizedDescription,
+            @"helpAnchor": error.helpAnchor ?: [NSNull null],
+            @"failureReason": error.localizedFailureReason ?: [NSNull null],
+            @"recoverySuggestion": error.localizedRecoverySuggestion ?: [NSNull null]
+        }
+      });
+    } else if (image && strongSelf.onLoad) {
+      strongSelf.onLoad(@{
+        @"cacheType": @([self convertToCacheTypeEnum:cacheType]),
+        @"source": @{
+            @"url": imageURL.absoluteString,
+            @"width": @(image.size.width),
+            @"height": @(image.size.height),
+            @"mediaType": [self sdImageFormatToMediaType:image.sd_imageFormat] ?: [NSNull null]
+        }
+      });
+    }
+
+    if (strongSelf.onLoadEnd) {
+      strongSelf.onLoadEnd(@{});
+    }
+  };
+}
+
+- (EXImageCacheTypeEnum)convertToCacheTypeEnum:(SDImageCacheType)imageCacheType
+{
+  switch (imageCacheType) {
```

Comment:
I was wondering what `SDImageCacheTypeAll`'s description here means:

https://github.com/SDWebImage/SDWebImage/blob/09f06159a3284f6981d5495728e5c3cb3dfb82fa/SDWebImage/Core/SDImageCacheDefine.h#L31-L35

I came to conclusion that this will never end up as `cacheType` of the downloaded image (what would it mean? That the image was present in both memory and disk but we decided to fetch from disk?) That's why I decided to return `unknown` here in that case (which we should never hit). What is your opinion?

Reply:
I think you're right. Looking a the SDWebImage code, the `SDWebImageCacheTypeAll` doesn't seem to be used when retrieving data, but only when storing/clearing data to the cache. 👍 

This shouldn't happen, but perhaps we can add some comments to the code, briefly explaining the case.